File: models/Gpt-3.5-turbo/cpp/code/problems/graph_algorithms/Cycle_Finding.cpp

#include <iostream>
#include <vector>

using namespace std;

struct Edge {
    int from, to, weight;
};

const int INF = 1e9;

vector<int> bellmanFord(int n, vector<Edge>& edges) {
    vector<int> dist(n + 1, INF);
    vector<int> parent(n + 1, -1);

    int x;
    for (int i = 0; i < n; ++i) {
        x = -1;
        for (Edge& edge : edges) {
            if (dist[edge.from] + edge.weight < dist[edge.to]) {
                dist[edge.to] = max(-INF, dist[edge.from] + edge.weight);
                parent[edge.to] = edge.from;
                x = edge.to;
            }
        }
    }

    if (x == -1) {
        return {};
    }

    for (int i = 0; i < n; ++i) {
        x = parent[x];
    }

    vector<int> cycle;
    for (int v = x;; v = parent[v]) {
        cycle.push_back(v);
        if (v == x && cycle.size() > 1) {
            break;
        }
    }

    reverse(cycle.begin(), cycle.end());
    return cycle;
}

int main() {
    int n, m;
    cin >> n >> m;

    vector<Edge> edges(m);
    for (int i = 0; i < m; ++i) {
        cin >> edges[i].from >> edges[i].to >> edges[i].weight;
    }

    vector<int> cycle = bellmanFord(n, edges);
    if (cycle.empty()) {
        cout << "NO" << endl;
    } else {
        cout << "YES" << endl;
        for (int node : cycle) {
            cout << node << " ";
        }
        cout << endl;
    }

    return 0;
}

File: models/Gpt-3.5-turbo/py/code/graph_algorithms/Cycle_Finding.py

def find_negative_cycle(n, m, edges):
    dist = [float('inf')] * (n + 1)
    parent = [-1] * (n + 1)
    cycle = []

    # Step 1: Set the distance of the source node to 0
    dist[1] = 0

    # Step 2: Relax each edge for n-1 times
    for _ in range(n - 1):
        for a, b, c in edges:
            if dist[a] + c < dist[b]:
                dist[b] = dist[a] + c
                parent[b] = a

    # Step 3: Check for negative cycles
    for a, b, c in edges:
        if dist[a] + c < dist[b]:
            # Step 4: Find negative cycle
            x = b
            for _ in range(n):
                x = parent[x]
            y = x
            while True:
                cycle.append(y)
                if y == x and len(cycle) > 1:
                    break
                y = parent[y]
            cycle.reverse()
            return "YES", cycle

    # Step 5: No negative cycle found
    return "NO"

# Read input
n, m = map(int, input().split())
edges = []
for _ in range(m):
    a, b, c = map(int, input().split())
    edges.append((a, b, c))

# Find negative cycle
result = find_negative_cycle(n, m, edges)

# Print output
print(result[0])
if result[0] == "YES":
    print(' '.join(map(str, result[1])))
